File: java/BinarySearch/Magnetic_Force_Between_Two_Balls_1552.java
package java.BinarySearch;

import java.util.*;
import java.io.*;

// import static java.lang.Math.max;
// import static java.lang.Math.min;
// import static java.lang.Math.abs;
// import static java.lang.System.out;

public class Magnetic_Force_Between_Two_Balls_1552 {
    /*
     * same as Split array Largest sum
     */
    public int maxDistance(int[] position, int m) {
        Arrays.sort(position); // sort is important
        int start = 1;
        int end = position[position.length - 1];
        int mid, count, prev;
        while (start <= end) {
            mid = start + (end - start) / 2;
            count = 1;
            prev = position[0];
            for (int i = 1; i < position.length; ++i) {
                if (position[i] - prev >= mid) {
                    count++;
                    prev = position[i];
                }
                if (count >= m)
                    break;
            }
            if (count >= m) {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }
        return end;
    }

    public static void main(String[] args) throws FileNotFoundException {
        File f = new File("output.txt");

        // Scanner ip = new Scanner(System.in);
        try (Scanner ip = new Scanner(new File("input.txt"));
                PrintWriter op = new PrintWriter(f);) {
            Magnetic_Force_Between_Two_Balls_1552 obj = new Magnetic_Force_Between_Two_Balls_1552();
            int nums[] = { 1, 2, 3, 4, 7 };
            // Input: position = [5,4,3,2,1,1000000000], m = 2 another one
            op.println(obj.maxDistance(nums, 3));

        }
    }

}

File: Code/BinarySearch/Magnetic_Force_Between_Two_Balls_1552.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    /*
       same as Split array bw largest sum
     */
    int maxDistance(vector<int> position, int m)
    {
        sort(position.begin(), position.end()); // sort is important
        int start = 1;
        int end = position[position.size() - 1];
        int mid, count, prev;
        while (start <= end)
        {
            mid = start + (end - start) / 2;
            count = 1;
            prev = position[0];
            for (int i = 1; i < position.size(); ++i)
            {
                if (position[i] - prev >= mid)
                {
                    count++;
                    prev = position[i];
                }
            }
            if (count >= m)
            {
                start = mid + 1;
            }
            else
            {
                end = mid - 1;
            }
        }
        return end;
    }
};
int main()
{
#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif
    Solution obj;
    auto ans = obj.maxDistance({79, 74, 57, 22}, 4);
    cout << ans;
    return 0;
}